### PR TITLE
[ENH] Visualization parameters

### PR DESCRIPTION
**Pull request by markotoplak**

### CODE REVIEW

**Review by stuart-cls — COMMENTED** [review]
Very nice! I have a few comments:

* Widgets that re-use the Spectra widget (Peakfit, Hyperspectra) are losing the ability to annotate the figures. I guess we can plan to re-add it but we should make an issue to track that
* How did you decide which visual items to put here and which ones to leave in the little menu? It seems arbitrary. It's a much nicer experience typing into the separate window and seeing your changes updated on the graph than trying to edit things one-by-one in the little drop-down. In particular, I'm thinking of the **Define view range**, which I find currently hard to use.

Otherwise, I didn't see any problems bringing old workflows forward into this.

Thanks,
Stuart

File: orangecontrib/spectroscopy/tests/test_owspectra.py [hunk 1]
```diff
@@ -554,3 +555,91 @@ def test_compat_no_group(self):
         self.assertTrue(self.widget.compat_no_group)
         # We decided to remove the info box
         # self.assertTrue(self.widget.Information.compat_no_group.is_shown())
+
+    def test_visual_settings(self, timeout=5):
+        graph = self.widget.curveplot
+        font = QFont()
+        font.setItalic(True)
+        font.setFamily("Helvetica")
```
[the remark this answers is not in the document]
Comment (borondics) [reply orangecontrib/spectroscopy/tests/test_owspectra.py:563]: Hi @stuart-cls,

Thanks for checking this! I agree, we should add all visual features to this new window from the current menu.

@markotoplak said that there is a way to handle two graph windows and then hopefully these issues will be nicely handled.

For the font list I also get some crazy stuff - not sure how to limit these well, maybe allow only a preset list of fonts and check if those are installed?

[the remark this answers is not in the document]
Comment (markotoplak) [reply orangecontrib/spectroscopy/tests/test_owspectra.py:563]: @stuart-cls, I do not quite understand, does that test thus fail?

Yes, each system has a different set of installed fonts, which is a problem. We discussed available fonts some time ago in Orange3, here: https://github.com/biolab/orange3/pull/4892

File: orangecontrib/spectroscopy/widgets/owspectra.py [hunk 2]
```diff
@@ -77,6 +79,50 @@ def selection_modifiers():
...
+class ParameterSetter(CommonParameterSetter):
+    MEAN_LABEL = "Mean"
+    LINE_LABEL = "Lines"
+    SEL_LINE_LABEL = "Selected lines"
+    RANGE_LABEL = "Range"
+    SEL_RANGE_LABEL = "Selected range"
+
+    def __init__(self, master):
+        super().__init__()
+        self.master = master # how does this come here?
```
[the remark this answers is not in the document]
Comment (markotoplak) [reply orangecontrib/spectroscopy/widgets/owspectra.py:91]: Hahaha, those are Feri's notes. Need to delete them. Good catch, thanks!

**Comment by borondics** [comment]
@markotoplak is there any chance we could merge this in the next few days? I am not sure if we have the line width changing feature available also, but I would also need that.

**Comment by markotoplak** [comment]
I added the same basic visualization parameters to the HyperSpectra widget - mostly to see if I can use a single settings window to control both visualizations.

So now we need to decide where to go. I'd need advice:
- What should we add to these settings? I think adding axis limits (define view range) would make sense, also we could add line thickness, which is now not a parameter. Any other ideas?
- Where do we add them to? Do we also need these for preprocessor widgets? I believe visualizations in those widgets are not meant to become final plots so I am leaning towards "no".

And ideas? @borondics, @stuart-cls

**Comment by stuart-cls** [comment]
Very nice!

I agree, axis limits and line widths should be included (line widths is a frequent request).

For widgets, I agree about Preprocessor. I think Spectral Series should have it. I'm on the fence about Peak Fit: people want to use the fits plots but no one likes how awkward the current implementation makes that :) (it was only meant to show a representative few fits)

**Comment by borondics** [comment]
Cool stuff!! Thanks :)

I was wondering about the single parameter window possibility. The problem there will be that 1D spectra and 2D map plots require slightly different stuff that can mess up a bit the interface. I wonder if it would be nicer to have the visual settings belong to the axes and not the widget. Or somehow make two little icons in the widget status bar? Or just have one icon that would give a dropdown for the individual plots?

I agree with those parameters (display limits and line width) and there is one thing I've been dreaming about and I think it prevents us using the plots in paper: figure sizing. I know that the axes sizing is hard (although that would be ideal) but we could at least try something with the window that would make the axes area always the same?

**Comment by stuart-cls** [comment]
I think the combined settings for both figures in Hyperspectra works well and is intuitive. I can't imagine a situation where you need different Title formats for the two figures.

Good point about figure sizing / aspect ratio. That stuff would fit nicely in this menu as well if feasible.

I reviewed what the interactive Matplotlib window exposes as GUI settings, and we've basically covered it all if we include figure sizing.

**Comment by markotoplak** [comment]
Inspired by wrangling visual settings with @borondics I added view range limits for spectra (where we already had them in menus). Because they work differently that they used to - whetever is typed in is strictly adhered to in all (I hope) situations - I could not reuse the old settings (the old behaviour was that a user had to apply them manually, they were just saved).

I think this is not ready to be merged. Other settings can (and therefore should) be added separately.

Can anyone test it out a bit?

**Comment by stuart-cls** [comment]
Thanks Marko, this is nice! Having the separate window to live edit the figure is a great experience.

Regarding the View Range, it's quite weird now. When the View Range is active, none of the interactive figure tools work (zoom, rescale, etc). I guess this is the hard trade-off of optimizing for a particular figure, and exploring data, but this seems like too much of a swing toward the former. 

A possible solution is to have some visible indicator (I think pyqtgraph does this with a little "A" button for autoscaling) that shows when the fixed limits are applied vs when you are zooming around. We could add (another!) setting in the drop down to toggle between them. That would at least be explicit, which is better than the old way where it was hard to tell if the figure limits were currently applied or not. The dynamic zoom tools could then be disabled in fixed limits mode.

One last thing, the "Reset" button. It's great to have a global reset but I was quite surprised to lose my figure captions when all I wanted was to reset the View Range limits! I predict this to be a future source of hair-pulling by our users.

**Comment by markotoplak** [comment]
> Regarding the View Range, it's quite weird now. When the View Range is active, none of the interactive figure tools work (zoom, rescale, etc). I guess this is the hard trade-off of optimizing for a particular figure, and exploring data, but this seems like too much of a swing toward the former.

Yes, I know. I tried but could not think of a good compromise of actually setting view range limits and being able to explore, so then opted for two extremes: (1) no limits, free graph, explore as you will (2) limits, strict for export/display just as the user wants. 

> We could add (another!) setting in the drop down to toggle between them. That would at least be explicit, which is better than the old way where it was hard to tell if the figure limits were currently applied or not. The dynamic zoom tools could then be disabled in fixed limits mode.

So, if I simplify, you suggest adding a setting whether to observe set limits or not? I do agree this is easy to add, but I am not sure it is needed. I am limited by the dialog here. What we could do is add a checkbox in this dialog that enables/disables limits, if this is a common-enough operation.

I would not want to put that setting into the menu, because then you have two separate places for something very related and could be confusing (and would then really need some visual indicators, but I doubt even that would be enough to make it clear).


> One last thing, the "Reset" button. It's great to have a global reset but I was quite surprised to lose my figure captions when all I wanted was to reset the View Range limits! I predict this to be a future source of hair-pulling by our users.

Funny, I did not notice that Reset button before. Yeah, I understand what can happen.. I hope for a single hair-splitting session per user. Perhaps we should add some space between settings and the buttons part... No idea what is the best thing to do there.

**Comment by markotoplak** [comment]
I have added an Information that axes are locked. Before that, I tried adding a lock symbol on the graph, but positioning it proved tedious, plus it did not seem that clear.

**Pull request merged**